File: p1/p198/Rob.java
package leetcode.p1.p198;

public class Rob {
    /**
     * 1D DP
     * Time complexity: O(n)
     * Space complexity: O(n)
     */
    public int rob1(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        int[] dp = new int[nums.length];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[1], dp[0]);
        for (int i = 2; i < dp.length; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);
        }
        return dp[dp.length - 1];
    }


    /**
     * DP
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        int[] dp = new int[2];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[1], dp[0]);
        for (int i = 2; i < nums.length; i++) {
            int temp = Math.max(dp[1], dp[0] + nums[i]);
            dp[0] = dp[1];
            dp[1] = temp;
        }
        return dp[1];
    }
}

File: p1/p198/rob.py
from typing import List


class Solution:
    """
    Feb 15, 2025 13:25
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        dp = [nums[0], max(nums[1], nums[0])]

        for i in range(2, len(nums)):
            temp = max(dp[0] + nums[i], dp[1])
            dp[0] = dp[1]
            dp[1] = temp

        return max(dp)


if __name__ == '__main__':
    obj = Solution()
    print(obj.rob([2, 1, 1, 2]))
